src/Compilers/CSharp/Portable/Parser/Lexer.cs

```diff
@@ -2381,8 +2403,12 @@ private CSharpSyntaxNode LexSingleDirective(
 
             directive = _directiveParser.ParseDirective(isActive, endIsActive, afterFirstToken, afterNonWhitespaceOnLine);
 
-            this.AddTrivia(directive, ref triviaList);
-            _directives = directive.ApplyDirectives(_directives);
+            if (!afterNonWhitespaceOnLine)
```

Comment:
>if (!afterNonWhitespaceOnLine) [](http://example.com/codeflow?start=12&length=30)

I would like to suggest getting rid of the conditional logic around `afterNonWhitespaceOnLine` and stopping passing it around by refactoring the code a little bit:
1. Extract the code that produces ```directive``` above, plus ```_mode = saveMode;``` into a reusable helper with signature ```CSharpSyntaxNode ParseDirective(bool isActive, bool endIsActive, bool afterFirstToken)```. 
2. Call the helper here and from ```LexSyntaxTrivia``` when we are parsing the directive for error recovery.   
3. In ```LexSyntaxTrivia```, pass `false` for afterFirstToken to the helper. We need to decide what are the right values for ```isActive```, ```endIsActive```. Right now they both are implicitly `true` for that code path, but we need to make an explicit decision about each of them. We might end up using the same values, just need to understand why these are the right values for the case.
4. Get rid of the ```afterNonWhitespaceOnLine``` parameter from all the helpers and adjust their code as if its value was always false. This includes the target of ```_directiveParser.ParseDirective``` call.